docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,692 @@
+|---------------------|------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|
+| NULL FLAG           | `-3`       | This flag indicates the object is a null value. We don't use another byte to indicate REF, so that we can save one byte.                                |
+| REF FLAG            | `-2`       | This flag indicates the object is already serialized previously, and fury will write a ref id with unsigned varint format instead of serialize it again |
+| NOT_NULL VALUE FLAG | `-1`       | This flag indicates the object is a non-null value and fury doesn't track ref for this type of object.                                                  |
+| REF VALUE FLAG      | `0`        | This flag indicates the object is referencable and the first time to serialize.                                                                         |
+
+When reference tracking is disabled globally or for specific types, or for certain types within a particular
+context(e.g., a field of a type), only the `NULL` and `NOT_NULL VALUE` flags will be used for reference meta.
+
+For languages which doesn't support reference such as rust, reference tracking must be disabled for correct
+deserialization by fury rust implementation.
+
+For languages whose object values are not null by default:
+
+- In rust, Fury takes `Option:None` as a null value
+- In c++, Fury takes `std::nullopt` as a null value
+- In golang, Fury takes `null interface/pointer` as a null value
+
+If one want to deserialize in languages like `Java/Python/JavaScript`, he should mark the type with all fields
+not-null by default, or using schema-evolution mode to carry the not-null fields info in the data.
+
+## Type Meta
+
+For every type to be serialized, it must be registered with an optional ID first. The registered type will have a
+user-provided or an auto-growing unsigned int i.e. `type_id`. The registration can be used for security check and type
+identification. The id of user registered type will be added by `64` to make space for Fury internal data types.
+
+Depending on whether meta share mode and registration is enabled for current type, Fury will write type meta
+differently.
+
+### Schema consistent
+
+If schema consistent mode is enabled globally or enabled for current type, type meta will be written as a fury unsigned
+varint of `type_id`.
+
+### Schema evolution
+
+If schema evolution mode is enabled globally or enabled for current type, type meta will be written as follows:
+
+- If meta share mode is not enabled:
+    - If type meta hasn't been written before, the data will be written as:
+      ```
+      | unsigned varint: 0b11111111 | type def |
+      ```
+    - If type meta has been written before, the data will be written as:
+      ```
+      | unsigned varint: written index << 1 | type def |
+      ```
+- If meta share mode is enabled:
```

Comment:
The only difference I noticed between the two modes is that one mode sends metadata and the other mode does not.
If the share mode is not enabled, can we write the meta data at the `meta start offset` ?  When it is disabled, we leave the `meta start offset` empty. so we can more easily understand the difference between the two modes.

Reply:
update to :

- Normal mode(meta share not enabled):
    - If type meta hasn't been written before, add `type def`
      to `captured_type_defs`: `captured_type_defs[type def] = map size`
    - Get index of the meta in `captured_type_defs`, write that index as an unsigned varint.
    - The meta will be written as:
      ```
      | unsigned varint: written index |
      ```
    - After finished the serialization of the object graph, fury will start to write `captured_type_defs`:
        - Firstly, set current to `meta start offset` of fury header
        - Then write `captured_type_defs` one by one:
          ```python
          buffer.write_var_uint32(len(writting_type_defs))
          for type_meta in writting_type_defs:
          type_meta.write_type_def(buffer)
          ```
- Meta share mode: the writing steps are same as the normal mode, but `captured_type_defs` will be shared across
  multiple serializations of different objects. For example, suppose we have a batch to serialize:
    ```python
    captured_type_defs = {}
    stream = ...
    # add `Type1` to `captured_type_defs` and write `Type1`
    fury.serialize(stream, [Type1()])
    # add `Type2` to `captured_type_defs` and write `Type2`, `Type1` is written before.
    fury.serialize(stream, [Type1(), Type2()])
    # `Type1` and `Type2` are written before, no need to write meta.
    fury.serialize(stream, [Type1(), Type2()])
    ```

- Streaming mode(streaming mode doesn't support meta share):
    - If type meta hasn't been written before, the data will be written as:
      ```
      | unsigned varint: 0b11111111 | type def |
      ```
    - If type meta has been written before, the data will be written as:
      ```
      | unsigned varint: written index << 1 |
      ```
      `written index` is the id in `captured_type_defs`.
    - With this mode, `meta start offset` can be omitted.